src/main/scala/com/fulcrumgenomics/alignment/Aligner.scala

```diff
@@ -86,19 +85,86 @@ object Aligner {
+    def targetLength: Int
+  }
+
   /** A single alignment matrix for a given `Direction` storing both the scoring and traceback matrices produce by the aligner. */
-  case class AlignmentMatrix(direction: Direction, scoring: Matrix[Int], trace: Matrix[Direction]) {
+  private case class SimpleAlignmentMatrix(direction: Direction, scoring: Matrix[Int], trace: Matrix[Direction]) extends AlignmentMatrix {
     val queryLength: Int  = scoring.x - 1
     val targetLength: Int = scoring.y - 1
   }
 
+  object CachedAligner {
+
+    case class CachedAlignmentMatrix
+    (direction: Direction,
+     scoring: Matrix[Int],
+     trace: Matrix[Direction],
+     queryLength: Int,
+     targetLength: Int) extends AlignmentMatrix {
+      require(queryLength < scoring.x)
+      require(queryLength < trace.x)
+      require(targetLength < scoring.y)
+      require(targetLength < trace.y)
+    }
+
+    def apply(matchScore: Int, mismatchScore: Int, gapOpen: Int, gapExtend: Int, mode: Mode = Global): CachedAligner = {
+      val aligner = Aligner(matchScore, mismatchScore, gapOpen, gapExtend, mode)
+      new CachedAligner(scorer=aligner.scorer, mode=aligner.mode)
+    }
+  }
+
+  class CachedAligner(scorer: AlignmentScorer,
+                      useEqualsAndX: Boolean = true,
+                      mode: Mode = Global,
+                      initQueryLength: Int = 1024,
+                      initTargetLength: Int = 1024) extends Aligner(scorer=scorer, useEqualsAndX=useEqualsAndX, mode=mode) {
+    import CachedAligner.CachedAlignmentMatrix
+
+    private var matrices: Array[CachedAlignmentMatrix] = AllDirections.sorted.map { dir =>
+      CachedAlignmentMatrix(
+        direction    = dir,
+        scoring      = Matrix[Int](initQueryLength+1, initTargetLength+1),
+        trace        = Matrix[Direction](initQueryLength+1, initTargetLength+1),
+        queryLength  = initQueryLength,
+        targetLength = initTargetLength
+      )
+    }.toArray
+
+    override protected def allocMatrices(query: Array[Byte], target: Array[Byte]): Array[AlignmentMatrix] = {
+      if (query.length + 1 > this.matrices(0).scoring.x || target.length + 1 > this.matrices(0).scoring.y) {
+        val xLength = math.max(query.length + 1, this.matrices(0).scoring.x)
+        val yLength = math.max(target.length + 1, this.matrices(0).scoring.y)
+        this.matrices = AllDirections.sorted.map { dir =>
+          CachedAlignmentMatrix(
+            direction    = dir,
+            scoring      = Matrix[Int](xLength, yLength),
+            trace        = Matrix[Direction](xLength, yLength),
+            queryLength  = query.length,
+            targetLength = target.length
+          )
+        }.toArray
+      } else if (this.matrices(0).queryLength != query.length || this.matrices(0).targetLength != target.length) {
+        this.matrices = this.matrices.map { matrix =>
+          matrix.copy(queryLength=query.length, targetLength=target.length)
+        }
```

Comment:
This will leave the matrices `scoring` and `trace` with different sizes than the stated `queryLength` and `targetLength`. Is that OK?